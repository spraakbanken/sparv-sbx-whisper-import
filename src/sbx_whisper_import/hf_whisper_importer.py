"""Whisper importer using transformers."""

import typing as t

import torch
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline


class TranscribeChunk(t.TypedDict):
    """Chunked text with timestamp (start, end)."""

    timestamp: tuple[float, float]
    text: str


class TranscribeResult(t.TypedDict):
    """Result from transcribe."""

    text: str
    chunks: list[TranscribeChunk]


class HFWhisperImporter:
    def __init__(self, *, model_id: str) -> None:
        device = "cuda:0" if torch.cuda.is_available() else "cpu"
        torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

        model = AutoModelForSpeechSeq2Seq.from_pretrained(
            model_id,
            torch_dtype=torch_dtype,
            use_safetensors=True,
        )
        model.to(device)
        processor = AutoProcessor.from_pretrained(model_id)

        pipe = pipeline(
            "automatic-speech-recognition",
            model=model,
            tokenizer=processor.tokenizer,
            feature_extractor=processor.feature_extractor,
            torch_dtype=torch_dtype,
            device=device,
        )

        generate_kwargs = {
            "task": "transcribe",
            "language": "sv",
        }

        self._pipe = pipe
        self._generate_kwargs = generate_kwargs

    def transcribe(self, audio_path: str) -> TranscribeResult:
        """Transcribe audio from given path.

        Args:
            audio_path: A path to the audio to transcribe.

        Returns:
            The transcribed text along with chunks.
        """
        return self._pipe(audio_path, chunk_length_s=30, generate_kwargs=self._generate_kwargs, return_timestamps=True)


if __name__ == "__main__":
    import sys

    importer = HFWhisperImporter(model_id="KBLab/kb-whisper-tiny")
    audio_path = sys.argv[1]

    res = importer.transcribe(audio_path)
    print(res)
