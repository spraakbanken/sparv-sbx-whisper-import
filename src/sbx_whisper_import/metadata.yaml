id: sbx-swe-speech2text-transformers-kb_whisper
languages_codes:
  - swe
keywords:
  - whisper
  - transformers
  - huggingface
tools:
  - name: "transformers"
    url: "https://github.com/huggingface/transformers"
    license: "Apache-2.0 license"
models:
  - name: "KBLab/kb-whisper-tiny"
    url: "https:/huggingface.co/KBLab/kb-whisper-tiny"
    description: |-
      The National Library of Sweden releases a new suite of Whisper models trained
      on over 50,000 hours of Swedish speech.
    license: "Apache-2.0 license"
other_references:
  - "Swedish Whispers; Leveraging a Massive Speech Corpus for Swedish Speech Recognition - Vesterbacka et al - 2025 - Interspeech 2025, p758-762 - doi: 10.21437/Interspeech.2025-2646"
short_description:
  swe: Transkription av ljudfil till text med transformers och KB-Whisper för svenska
  eng: Transcribing audio file to text by transformers and KB-Whisper for Swedish
created: 2025-09-04
example_output: |-
  ```xml
  <?xml version='1.0' encoding='utf-8'?>
  <text source_filename="example.mp3">
    <utterance end="6.0" start="0.0">Världsförklaring</utterance>
  </text>
  ```
plugin_url: https://github.com/spraakbanken/sparv-sbx-whisper-import
type: utility
license: MIT license
sparv_handler: sbx_whisper_import:parse_mp3
