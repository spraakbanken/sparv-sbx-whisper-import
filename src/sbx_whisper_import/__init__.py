"""Sparv plugin to import audio files as text with Whisper."""

from sparv.api import Config

from sbx_whisper_import.whisper_import import parse_mp3, parse_wav

__all__ = ["parse_mp3", "parse_wav"]

__config__ = [
    Config(
        "sbx_whisper_import.model_size",
        "small",
        description="The size of the model. Defaults to 'small'",
        datatype=str,
    ),
    Config(
        "sbx_whisper_import.model_verbosity",
        "standard",
        description="The verbosity of the model. Defaults to 'standard'",
        datatype=str,
    ),
]

__description__ = "Import audio files as text with Whisper."

__version__ = "0.1.0"
