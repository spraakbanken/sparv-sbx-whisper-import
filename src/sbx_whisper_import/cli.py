"""Command-line interface for sbx-whisper-cli."""

import argparse
import json
import typing as t

from sbx_whisper_import.hf_whisper_importer import HFWhisperImporter, TranscribeResult


def main() -> None:
    """Run sbx-whisper-import as cli."""
    res = _run_transcription()
    print(json.dumps(res, ensure_ascii=False))  # noqa: T201


def _run_transcription(argv: t.Sequence[str] | None = None) -> TranscribeResult:
    parser = _build_cli()

    args = parser.parse_args(argv)

    importer = HFWhisperImporter(model_size=args.model_size, model_verbosity=args.verbosity, verbose=not args.quiet)

    return importer.transcribe(args.input)


def _build_cli() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Transcribe audio file with KB-Whisper. Output is in JSON.")
    parser.add_argument("--model-size", type=str, default="small", help="set the size of the model")
    parser.add_argument("--verbosity", type=str, default="standard", help="set the verbosity of the model")
    parser.add_argument(
        "input", type=str, help="audio input to trancribe in one of the formats MP3, OGG or WAV", metavar="INPUT"
    )
    parser.add_argument("-q", "--quiet", action="store_true", help="Silence output", default=False)
    return parser


if __name__ == "__main__":
    main()
