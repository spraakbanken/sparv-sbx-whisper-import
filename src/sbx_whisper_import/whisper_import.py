"""Sparv importer for using Whisper."""

from sparv import api as sparv_api
from sparv.api import Config, Output, Source, SourceFilename, SourceStructure, Text

from sbx_whisper_import.hf_whisper_importer import HFWhisperImporter

logger = sparv_api.get_logger(__name__)


@sparv_api.importer(
    "Import audio from MP3 with Whisper",
    file_extension="mp3",
    outputs=["text"],
    text_annotation="text",
    config=[
        Config(
            "sbx_whisper_import.model_size",
            "small",
            description="The size of the model. Defaults to 'small'",
            datatype=str,
        ),
        Config(
            "sbx_whisper_import.model_verbosity",
            "standard",
            description="The verbosity of the model. Defaults to 'standard'",
            datatype=str,
        ),
    ],
)
def parse_mp3(
    source_file: SourceFilename = SourceFilename(),
    source_dir: Source = Source(),
    model_size: str = Config("sbx_whisper_import.model_size"),
    model_verbosity: str = Config("sbx_whisper_import.model_verbosity"),
) -> None:
    """Transcribe mp3 file as input to Sparv."""
    transcribe_audio(
        source_file=source_file,
        source_dir=source_dir,
        model_size=model_size,
        model_verbosity=model_verbosity,
        extension=".mp3",
    )


def transcribe_audio(
    source_file: SourceFilename,
    source_dir: Source,
    model_size: str,
    model_verbosity: str,
    extension: str,
) -> None:
    """Transcribe audio file as input to Sparv."""
    importer = HFWhisperImporter(model_size=model_size, model_verbosity=model_verbosity)

    res = importer.transcribe(str(source_dir.get_path(source_file, extension)))

    logger.debug("res=%s", res)

    Text(source_file).write(res["text"])

    utterance_spans: list[tuple[int, int]] = []
    utterance_starts: list[float] = []
    utterance_ends: list[float] = []
    curr_start: int = 0
    curr_end: int = 0
    for chunk in res["chunks"]:
        utterance_starts.append(chunk["timestamp"][0])
        utterance_ends.append(chunk["timestamp"][1])
        curr_end += len(chunk["text"])
        utterance_spans.append((curr_start, curr_end))
        curr_start = curr_end

    # Make up a text annotation surrounding the whole file
    text_annotation = ["text", "text:source_filename", "utterance", "utterance:start", "utterance:end"]
    source_file_name = f"{source_file}{extension}"
    Output("text", source_file=source_file).write([(0, len(res["text"]))])
    Output("text:source_filename", source_file=source_file).write([source_file_name])
    Output("utterance", source_file=source_file).write(utterance_spans)
    Output("utterance:start", source_file=source_file).write(utterance_starts)
    Output("utterance:end", source_file=source_file).write(utterance_ends)
    SourceStructure(source_file).write(text_annotation)
