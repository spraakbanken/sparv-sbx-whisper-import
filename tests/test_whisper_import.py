from syrupy.assertion import SnapshotAssertion

from sbx_whisper_import import whisper_import


def test_transcribe_and_prepare_spans(snapshot: SnapshotAssertion) -> None:
    text, spans, starts, ends = whisper_import._transcribe_and_prepare_spans(
        "tiny", "standard", "assets/aspenstrom_varldsforklaring_aspenstrom.mp3"
    )

    assert text == snapshot
    assert spans == snapshot
    assert starts == snapshot
    assert ends == snapshot
