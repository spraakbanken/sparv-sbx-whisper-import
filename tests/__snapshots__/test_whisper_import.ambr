# serializer version: 1
# name: test_transcribe_and_prepare_spans
  ' Världsförklaring. När jag kommer ut på trappan har jag solen i brösthöjd. Nog är det en morgon i juli, men en morgon i juli är det inte. Vintergatan har i natt passerat genom trädgården, försilvrat mina dillkronor och kvarlämnat en stjärna i varje DAC-kåpa. Hallis Khomeet har fastnat i en enbuske. Jag stöder mig på vetenskapenkapen när jag påstår Det är oförklarligt och oförklarligt i dess toftlösa, interstellära Sjunger en fågel som jag inte ser och vars namn jag inte känner Inte finns det himmelsdag, himmelsväggar. Men himmelsdag, himmelsväggar finns det nog som återskallar, återskänker. Jag har sett detta ljus förr.'
# ---
# name: test_transcribe_and_prepare_spans.1
  list([
    tuple(
      0,
      18,
    ),
    tuple(
      18,
      74,
    ),
    tuple(
      74,
      137,
    ),
    tuple(
      137,
      187,
    ),
    tuple(
      187,
      258,
    ),
    tuple(
      258,
      299,
    ),
    tuple(
      299,
      349,
    ),
    tuple(
      349,
      386,
    ),
    tuple(
      386,
      417,
    ),
    tuple(
      417,
      451,
    ),
    tuple(
      451,
      481,
    ),
    tuple(
      481,
      523,
    ),
    tuple(
      523,
      597,
    ),
    tuple(
      597,
      627,
    ),
  ])
# ---
# name: test_transcribe_and_prepare_spans.2
  list([
    0.0,
    11.0,
    18.0,
    27.16,
    32.02,
    38.68,
    44.06,
    47.88,
    52.32,
    56.44,
    60.52,
    64.76,
    68.7,
    78.16,
  ])
# ---
# name: test_transcribe_and_prepare_spans.3
  list([
    6.0,
    18.0,
    25.62,
    31.76,
    37.92,
    41.24,
    47.72,
    52.16,
    56.28,
    60.36,
    64.6,
    68.18,
    77.14,
    81.5,
  ])
# ---
